arbor/mechcat.cpp

```diff
@@ -72,42 +71,26 @@ template <typename V>
 using string_map = std::unordered_map<std::string, V>;
 
 template <typename T>
-struct hopefully_typemap {
-    using type = util::either<T, std::exception_ptr>;
-};
-
-template <>
-struct hopefully_typemap<void> {
-    struct placeholder_type {};
-    using type = util::either<placeholder_type, std::exception_ptr>;
-};
-
-template <typename T>
-using hopefully = typename hopefully_typemap<T>::type;
+using hopefully = util::expected<T, std::exception_ptr>;
 
+namespace {
 // Convert hopefully<T> to T or throw.
-
 template <typename T>
-const T& value(const util::either<T, std::exception_ptr>& x) {
-    if (!x) {
-        std::rethrow_exception(x.second());
-    }
-    return x.first();
+const T& value(const hopefully<T>& x) {
+    if (!x) std::rethrow_exception(x.error());
+    return x.value();
 }
 
 template <typename T>
-T value(util::either<T, std::exception_ptr>&& x) {
-    if (!x) {
-        std::rethrow_exception(x.second());
-    }
-    return std::move(x.first());
+T value(hopefully<T>&& x) {
+    if (!x) std::rethrow_exception(std::move(x).error());
+    return std::move(x).value();
 }
 
-void value(const hopefully<void>& x) {
-    if (!x) {
-        std::rethrow_exception(x.second());
-    }
-}
+// Conveniently make an unexpected exception_ptr:
+template <typename X>
+auto unexpected_exception_ptr(X x) { return unexpected(make_exception_ptr(std::move(x))); }
```

Comment:
I see that you started to run out of vertical space during the refactoring ;)

Reply:
Also, `std::make_exception_ptr` is just bad naming, because it does a lot more than the usual `make_x` construction helper functions like `make_unique_ptr`.

Reply:
It was just getting a bit verbose :)